internal/corerad/advertise.go

```diff
@@ -96,12 +107,50 @@ type request struct {
 // and advertisement handling. Advertise will block until ctx is canceled or an
 // error occurs.
 func (a *Advertiser) Advertise(ctx context.Context) error {
-	// TODO: wait for readiness.
+	// Attempt immediate initialization and fall back to reinit loop if that
+	// does not succeed.
 	autoPrev, err := a.init()
 	if err != nil {
-		return fmt.Errorf("failed to wait for initialize %q advertiser: %v", a.ifi.Name, err)
+		autoPrev, err = a.reinit(ctx, err)
+		if err != nil {
+			// Don't block user shutdown.
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
+
+			return fmt.Errorf("failed to initialize %q advertiser: %w", a.iface, err)
+		}
 	}
 
+	for {
+		err := a.advertise(ctx)
+		switch {
+		case errors.Is(err, context.Canceled):
+			// Intentional shutdown.
+			return a.shutdown(autoPrev)
+		case err == nil:
+			panic("corerad: advertise must never return nil error")
+		}
+
+		// We encountered an error. Try to reinitialize the Advertiser based
+		// on whether or not the error is deemed recoverable.
+		a.logf("error advertising, attempting to reinitialize")
+
+		autoPrev, err = a.reinit(ctx, err)
+		if err != nil {
+			// Don't block user shutdown.
+			if errors.Is(err, context.Canceled) {
+				return nil
+			}
+
+			return fmt.Errorf("failed to reinitialize %q advertiser: %v", a.iface, err)
```

Comment:
I was expecting some kind of netlink dynamicism here - e.g. if I delete and recreate a tun interface with the same name.

That's a bigger chunk of work to make happen though, so maybe I'm just being too SRE in figuring out ways to break stuff :)

Reply:
Can you clarify what you mean? As in, I should be able to recover from an interface with the same name being added and deleted?

I do intend to add a rtnetlink state watcher that will force a reinit on any link state change, but that's going to be an extra chunk of work, and I just wanted to lay the groundwork for doing that for now.